api/schedule/create.go

```diff
@@ -0,0 +1,230 @@
+
+	// capture body from API request
+	input := new(library.Schedule)
+
+	err := c.Bind(input)
+	if err != nil {
+		retErr := fmt.Errorf("unable to decode JSON for new schedule: %w", err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure the entry is valid
+	err = validateEntry(minimumFrequency, input.GetEntry())
+	if err != nil {
+		retErr := fmt.Errorf("schedule of %s is invalid: %w", input.GetName(), err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// ensure schedule name is defined
+	if input.GetName() == "" {
+		util.HandleError(c, http.StatusBadRequest, fmt.Errorf("schedule name must be set"))
+	}
+
+	// update engine logger with API metadata
+	//
+	// https://pkg.go.dev/github.com/sirupsen/logrus?tab=doc#Entry.WithFields
+	logrus.WithFields(logrus.Fields{
+		"org":  r.GetOrg(),
+		"repo": r.GetName(),
+		"user": u.GetName(),
+	}).Infof("creating new schedule %s", input.GetName())
+
+	// ensure repo is allowed to create new schedules
+	if !util.CheckAllowlist(r, allowlist) {
+		retErr := fmt.Errorf("unable to create schedule %s: %s is not on allowlist", input.GetName(), r.GetFullName())
+
+		util.HandleError(c, http.StatusForbidden, retErr)
+
+		return
+	}
+
+	s := new(library.Schedule)
+
+	// update fields in schedule object
+	s.SetCreatedBy(u.GetName())
+	s.SetRepoID(r.GetID())
+	s.SetName(input.GetName())
+	s.SetEntry(input.GetEntry())
+	s.SetCreatedAt(time.Now().UTC().Unix())
+	s.SetUpdatedAt(time.Now().UTC().Unix())
+	s.SetUpdatedBy(u.GetName())
+
+	// set the active field based off the input provided
+	if input.Active == nil {
+		// default active field to true
+		s.SetActive(true)
+	} else {
+		s.SetActive(input.GetActive())
+	}
+
+	// send API call to capture the schedule from the database
+	dbSchedule, err := database.FromContext(c).GetScheduleForRepo(r, input.GetName())
+	if err == nil && dbSchedule.GetActive() {
+		retErr := fmt.Errorf("unable to create schedule: %s is already active", input.GetName())
+
+		util.HandleError(c, http.StatusConflict, retErr)
+
+		return
+	}
+
+	if !r.GetActive() {
+		retErr := fmt.Errorf("unable to create schedule: %s repo %s is disabled", input.GetName(), r.GetFullName())
+
+		util.HandleError(c, http.StatusConflict, retErr)
+
+		return
+	}
+
+	// if the schedule exists but is inactive
+	if !dbSchedule.GetActive() && input.GetActive() {
+		// update the user who created the schedule
+		dbSchedule.SetUpdatedBy(u.GetName())
+		// activate the schedule
+		dbSchedule.SetActive(true)
+
+		// send API call to update the schedule
+		err = database.FromContext(c).UpdateSchedule(dbSchedule)
+		if err != nil {
+			retErr := fmt.Errorf("unable to set schedule %s to active: %w", dbSchedule.GetName(), err)
+
+			util.HandleError(c, http.StatusInternalServerError, retErr)
+
+			return
+		}
+
+		// send API call to capture the updated schedule
+		s, _ = database.FromContext(c).GetScheduleForRepo(r, dbSchedule.GetName())
```

Comment:
I'll be honest that I didn't think about handling the error since so much of the existing code base does not... How about we open up a separate GitHub issue to track updating all of these database calls?

Reply:
FWIW -

We should plan to have the `Create*` and `Update*` funcs return the type so we don't need these calls

i.e. https://github.com/go-vela/server/pull/701